lib/sdam/topology.js

```diff
@@ -414,12 +420,46 @@ function selectServers(topology, selector, timeout, start, callback) {
     return callback(null, servers);
   }
 
-  const duration = calculateDurationInMs(process.hrtime(start));
-  if (duration > timeout) {
+  const duration = calculateDurationInMs(start);
+  if (duration >= timeout) {
     return callback(new MongoTimeoutError(`Server selection timed out after ${timeout} ms`));
   }
 
-  // TODO: loop this, add monitoring
+  const retrySelection = () => {
+    // ensure all server monitors are attempting monitoring immediately
+    topology.s.servers.forEach(server => server.monitor());
+
+    const iterationTimer = setTimeout(() => {
+      callback(new MongoTimeoutError('Server selection timed out due to monitoring'));
+    }, topology.s.minHeartbeatIntervalMS);
+
+    topology.once('topologyDescriptionChanged', () => {
```

Comment:
Is there anything here to ensure that, if the callback has already been called with an error, we won't call it here?

Reply:
I've read this through a few times and I can't see how the callback would be called twice after an error has called it back. Can you show me how that might happen?